File: 0224-basic-calculator/0224-basic-calculator.cpp
class Solution {
private:
    int dfs(int& i, string s, vector<int> res, int t, char op){
        while(i<s.length() && s[i]!=')'){
            if(s[i]==' ') {
                i+=1;
                continue;
            } else if(isdigit(s[i])){
                t = t*10 + (s[i]-'0');
                i+=1;
            } else if(s[i] == '+' || s[i]=='-'){
                if(t!=0){
                    if(op == '+'){
                        res.push_back(t);
                    } else res.push_back(-t);
                }
                op=s[i];
                t=0;
                i+=1;
            } else if(s[i] == '('){
                t=0;
                vector<int> temp;
                i+=1;
                int val = dfs(i, s, temp, t, '+');
                if(op == '+') res.push_back(val);
                else res.push_back(-val);
                i+=1;
            }
        }
        
        if(t){
            if(op == '+') res.push_back(t);
            else res.push_back(-t);
        }
        
        return accumulate(res.begin(), res.end(), 0);
    }
public:
    int calculate(string s) {
        s = "(" + s +")";
        int i=0;
        vector<int> res;
        int t=0;
        char op='+';
        return dfs(i, s, res, t, op);
    }
};

File: 0224-basic-calculator/0224-basic-calculator.ts

let i: number;
function dfs(s: string, d: number, op: string){
    let st: number[] = [];
    while(i<s.length && s[i]!=")"){
        if(s[i]==" "){
            i+=1;
            continue;
        } else if(!isNaN(parseInt(s[i]))){
            // digit
            d = d*10 + (s[i].charCodeAt(0) - "0".charCodeAt(0));
        } else if(s[i]=="("){
            i+=1;
            d=0;
            let temp = dfs(s, d, "+");
            
            if(op == "+") st.push(temp);
            else st.push(-temp);
            
        } else {
            if(op == "+") st.push(d);
            else st.push(-d);
            op = s[i];
            d=0;
        }
        i+=1;
    }
    
    if(op == "+") st.push(d);
    else st.push(-d);
    
    let ans=0;
    for(let x of st) ans+=x;
    return ans;
    
}


function calculate(s: string): number {
    i=0;
    let d=0;
    let op = "+";
    s="("+s+")";
    return dfs(s, d, op);
};